lib/epoch.ex

```diff
@@ -4,32 +4,71 @@ defmodule DateTimeParser.Epoch do
   @token_key :unix_epoch
   @max_subsecond_digits 6
 
-  def parse(%{"seconds" => raw_seconds, "subseconds" => raw_subseconds}) do
-    with {seconds, ""} <- Integer.parse(raw_seconds) do
-      parsed_epoch =
-        case(raw_subseconds) do
-          "" ->
-            {seconds, {0, 0}}
+  def parse(%{"sign" => sign, "seconds" => raw_seconds, "subseconds" => raw_subseconds}) do
+    is_negative = if sign == "-", do: true, else: false
+    has_subseconds = if raw_subseconds == "", do: false, else: true
 
-          raw_subseconds ->
-            {subseconds, ""} = Float.parse("0.#{raw_subseconds}")
-            microseconds = (subseconds * :math.pow(10, 6)) |> trunc()
+    with {:ok, seconds} <- parse_seconds(raw_seconds, is_negative, has_subseconds),
+         {:ok, subseconds} <- parse_subseconds(raw_subseconds, is_negative) do
+      {:ok, [{@token_key, {seconds, subseconds}}], nil, nil, nil, nil}
+    end
+  end
 
-            truncated_microseconds =
-              microseconds
-              |> Integer.digits()
-              |> Enum.take(@max_subsecond_digits)
-              |> Integer.undigits()
+  @spec parse_seconds(String.t(), boolean(), boolean()) :: {:ok, integer()}
+  defp parse_seconds(raw_seconds, is_negative, has_subseconds)
+
+  defp parse_seconds(raw_seconds, true, true) do
+    with {:ok, seconds} <- parse_seconds(raw_seconds, true, false) do
+      {:ok, seconds - 1}
+    end
+  end
 
-            number_of_subsecond_digits = min(String.length(raw_subseconds), @max_subsecond_digits)
+  defp parse_seconds(raw_seconds, true, false) do
+    with {:ok, seconds} <- parse_seconds(raw_seconds, false, false) do
+      {:ok, seconds * -1}
+    end
+  end
 
-            {seconds, {truncated_microseconds, number_of_subsecond_digits}}
+  defp parse_seconds(raw_seconds, false, _) do
+    with {seconds, ""} <- Integer.parse(raw_seconds) do
+      {:ok, seconds}
+    end
+  end
+
+  @spec parse_subseconds(String.t(), boolean()) :: {:ok, {integer(), integer()}}
+  defp parse_subseconds(raw_subseconds, is_negative)
+
+  defp parse_subseconds("", _), do: {:ok, {0, 0}}
+
+  defp parse_subseconds(raw_subseconds, true) do
+    with {:ok, {truncated_microseconds, number_of_subsecond_digits}} <-
+           parse_subseconds(raw_subseconds, false) do
+      negative_truncated_microseconds =
+        if truncated_microseconds > 0 do
+          ((1 * :math.pow(10, 6)) |> trunc()) - truncated_microseconds
```

Comment:
do you want to go ahead and round it instead of truncating?

Reply:
I don't think it ends up making a difference in this case since this always ends up being `1_000_000`. Both `trunc()` and `round()` help avoid it being in a weird format. I did the math to try to make it clearer what was going on instead of having a magic number.

I think https://github.com/taxjar/date_time_parser/pull/24/files#diff-0c651f0cef46a594e7e817decdccd3dcR59 is where the rounding would cause differences to happen.

Reply:
What do you think of 851eae3225dd77e9e09871b8fbf89b52d0a7c356?

Reply:
i like